File: src/brs/at/AT.java
/*
 * Some portion .. Copyright (c) 2014 CIYAM Developers

 Distributed under the MIT/X11 software license, please refer to the file LICENSE.txt
*/

package brs.at;


import brs.*;
import brs.db.BurstKey;
import brs.db.TransactionDb;
import brs.db.VersionedEntityTable;
import brs.services.AccountService;
import brs.util.Listener;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public class AT extends AtMachineState {

    private static final LinkedHashMap<Long, LinkedHashMap<Long, Long>> pendingFeesMap = new LinkedHashMap<>();
    private static final LinkedHashMap<Long, List<AtTransaction>> pendingTransactionsMap = new LinkedHashMap<>();
    public final BurstKey dbKey;
    private final String name;
    private final String description;
    private final int nextHeight;

    private AT(byte[] atId, byte[] creator, String name, String description, byte[] creationBytes, int height) {
        super(atId, creator, creationBytes, height);
        this.name = name;
        this.description = description;
        dbKey = atDbKeyFactory().newKey(AtApiHelper.getLong(atId));
        this.nextHeight = Burst.getBlockchain().getHeight();
    }

    public AT(byte[] atId, byte[] creator, String name, String description, short version,
              int height,
              byte[] stateBytes, int csize, int dsize, int cUserStackBytes, int cCallStackBytes,
              int creationBlockHeight, int sleepBetween, int nextHeight,
              boolean freezeWhenSameBalance, long minActivationAmount, byte[] apCode, long apCodeHashId) {
        super(atId, creator, version,
                height,
                stateBytes, csize, dsize, cUserStackBytes, cCallStackBytes,
                creationBlockHeight, sleepBetween,
                freezeWhenSameBalance, minActivationAmount, apCode, apCodeHashId);
        this.name = name;
        this.description = description;
        dbKey = atDbKeyFactory().newKey(AtApiHelper.getLong(atId));
        this.nextHeight = nextHeight;
    }

    public static void clearPendingFees(int blockHeight, long generatorId) {
        // using height+id as hash
        pendingFeesMap.remove(blockHeight + generatorId);
    }

    public static void clearPendingTransactions(int blockHeight, long generatorId) {
        pendingTransactionsMap.remove(blockHeight + generatorId);
    }

    public static void addPendingFee(long id, long fee, int blockHeight, long generatorId) {
        long hash = blockHeight + generatorId;
        LinkedHashMap<Long, Long> pendingFees = pendingFeesMap.get(hash);
        if(pendingFees == null) {
          pendingFees = new LinkedHashMap<>();
          pendingFeesMap.put(hash, pendingFees);
        }
        pendingFees.put(id, fee);
    }

    public static void addPendingFee(byte[] id, long fee, int blockHeight, long generatorId) {
        addPendingFee(AtApiHelper.getLong(id), fee, blockHeight, generatorId);
    }

    public static void addPendingTransaction(AtTransaction atTransaction, int blockHeight, long generatorId) {
        long hash = blockHeight + generatorId;
        List<AtTransaction> pendingTransactions = pendingTransactionsMap.get(hash);
        if(pendingTransactions == null) {
          pendingTransactions = new ArrayList<>();
          pendingTransactionsMap.put(hash, pendingTransactions);
        }
        pendingTransactions.add(atTransaction);
    }

    public static boolean findPendingTransaction(byte[] recipientId, int blockHeight, long generatorId) {
        long hash = blockHeight + generatorId;
        if(pendingTransactionsMap.get(hash) == null) {
          return false;
        }
        for (AtTransaction tx : pendingTransactionsMap.get(hash)) {
            if (Arrays.equals(recipientId, tx.getRecipientId())) {
                return true;
            }
        }
        return false;
    }

    private static BurstKey.LongKeyFactory<AT> atDbKeyFactory() {
        return Burst.getStores().getAtStore().getAtDbKeyFactory();
    }

    private static VersionedEntityTable<AT> atTable() {
        return Burst.getStores().getAtStore().getAtTable();
    }

    private static BurstKey.LongKeyFactory<ATState> atStateDbKeyFactory() {
        return Burst.getStores().getAtStore().getAtStateDbKeyFactory();
    }

    private static VersionedEntityTable<ATState> atStateTable() {
        return Burst.getStores().getAtStore().getAtStateTable();
    }

    public static AT getAT(byte[] id) {
        return getAT(AtApiHelper.getLong(id));
    }

    public static AT getAT(Long id) {
        return Burst.getStores().getAtStore().getAT(id, -1);
    }

    public static void addAT(Long atId, Long senderAccountId, String name, String description, byte[] creationBytes, int height) {
        ByteBuffer bf = ByteBuffer.allocate(8 + 8);
        bf.order(ByteOrder.LITTLE_ENDIAN);

        bf.putLong(atId);

        byte[] id = new byte[8];

        bf.putLong(8, senderAccountId);

        byte[] creator = new byte[8];
        bf.clear();
        bf.get(id, 0, 8);
        bf.get(creator, 0, 8);

        AT at = new AT(id, creator, name, description, creationBytes, height);

        AtController.resetMachine(at);

        atTable().insert(at);

        at.saveState();

        Account account = Account.getOrAddAccount(atId);
        account.apply(new byte[32], height);
    }

    public static List<Long> getOrderedATs() {
        return Burst.getStores().getAtStore().getOrderedATs();
    }

    public static byte[] compressState(byte[] stateBytes) {
        if (stateBytes == null || stateBytes.length == 0) {
            return null;
        }

        try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
            try (GZIPOutputStream gzip = new GZIPOutputStream(bos)) {
                gzip.write(stateBytes);
                gzip.flush();
            }
            return bos.toByteArray();
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    public static byte[] decompressState(byte[] stateBytes) {
        if (stateBytes == null || stateBytes.length == 0) {
            return null;
        }

        try (ByteArrayInputStream bis = new ByteArrayInputStream(stateBytes);
             GZIPInputStream gzip = new GZIPInputStream(bis);
             ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[256];
            int read;
            while ((read = gzip.read(buffer, 0, buffer.length)) > 0) {
                bos.write(buffer, 0, read);
            }
            bos.flush();
            return bos.toByteArray();
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    public void saveState() {
        ATState state = atStateTable().get(atStateDbKeyFactory().newKey(AtApiHelper.getLong(this.getId())));
        int prevHeight = Burst.getBlockchain().getHeight();
        int newNextHeight = prevHeight + getWaitForNumberOfBlocks();
        if (state != null) {
            state.setState(getState());
            state.setPrevHeight(prevHeight);
            state.setNextHeight(newNextHeight);
            state.setSleepBetween(getSleepBetween());
            state.setPrevBalance(getpBalance());
            state.setFreezeWhenSameBalance(freezeOnSameBalance());
            state.setMinActivationAmount(minActivationAmount());
        } else {
            state = new ATState(AtApiHelper.getLong(this.getId()),
                    getState(), newNextHeight, getSleepBetween(),
                    getpBalance(), freezeOnSameBalance(), minActivationAmount());
        }
        atStateTable().insert(state);
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public int nextHeight() {
        return nextHeight;
    }

    public static class HandleATBlockTransactionsListener implements Listener<Block> {
        private final AccountService accountService;
        private final Blockchain blockchain;
        private final TransactionDb transactionDb;

        public HandleATBlockTransactionsListener(AccountService accountService, Blockchain blockchain, TransactionDb transactionDb) {
            this.accountService = accountService;
            this.blockchain = blockchain;
            this.transactionDb = transactionDb;
        }

        @Override
        public void notify(Block block) {
            long hash = block.getHeight() + block.getGeneratorId();
            LinkedHashMap<Long, Long> pendingFees = pendingFeesMap.get(hash);
            if(pendingFees != null) {
              pendingFees.forEach((key, value) -> {
                Account atAccount = accountService.getAccount(key);
                accountService.addToBalanceAndUnconfirmedBalanceNQT(atAccount, -value);
              });
            }
            pendingFeesMap.remove(hash);

            List<Transaction> transactions = new ArrayList<>();
            List<AtTransaction> pendingTransactions = pendingTransactionsMap.get(hash);
            if(pendingTransactions != null) {
              for (AtTransaction atTransaction : pendingTransactions) {
                accountService.addToBalanceAndUnconfirmedBalanceNQT(accountService.getAccount(AtApiHelper.getLong(atTransaction.getSenderId())), -atTransaction.getAmount());
                accountService.addToBalanceAndUnconfirmedBalanceNQT(accountService.getOrAddAccount(AtApiHelper.getLong(atTransaction.getRecipientId())), atTransaction.getAmount());

                Transaction.Builder builder = new Transaction.Builder((byte) 1, Genesis.getCreatorPublicKey(),
                        atTransaction.getAmount(), 0L, block.getTimestamp(), (short) 1440, Attachment.AT_PAYMENT);

                builder.senderId(AtApiHelper.getLong(atTransaction.getSenderId()))
                        .recipientId(AtApiHelper.getLong(atTransaction.getRecipientId()))
                        .blockId(block.getId())
                        .height(block.getHeight())
                        .blockTimestamp(block.getTimestamp())
                        .ecBlockHeight(0)
                        .ecBlockId(0L);

                byte[] message = atTransaction.getMessage();
                if (message != null) {
                    builder.message(new Appendix.Message(message, blockchain.getHeight()));
                }

                try {
                    Transaction transaction = builder.build();
                    if (!transactionDb.hasTransaction(transaction.getId())) {
                        transactions.add(transaction);
                    }
                } catch (BurstException.NotValidException e) {
                    throw new RuntimeException("Failed to construct AT payment transaction", e);
                }
              }
              pendingTransactionsMap.remove(hash);
            }

            if (!transactions.isEmpty()) {
                // WATCH: Replace after transactions are converted!
                transactionDb.saveTransactions(transactions);
            }
        }
    }

    public static class ATState {

        public final BurstKey dbKey;
        private final long atId;
        private byte[] state;
        private int prevHeight;
        private int nextHeight;
        private int sleepBetween;
        private long prevBalance;
        private boolean freezeWhenSameBalance;
        private long minActivationAmount;

        protected ATState(long atId, byte[] state,
                          int nextHeight, int sleepBetween, long prevBalance, boolean freezeWhenSameBalance, long minActivationAmount) {
            this.atId = atId;
            this.dbKey = atStateDbKeyFactory().newKey(this.atId);
            this.state = state;
            this.nextHeight = nextHeight;
            this.sleepBetween = sleepBetween;
            this.prevBalance = prevBalance;
            this.freezeWhenSameBalance = freezeWhenSameBalance;
            this.minActivationAmount = minActivationAmount;
        }


        public long getATId() {
            return atId;
        }

        public byte[] getState() {
            return state;
        }

        void setState(byte[] newState) {
            state = newState;
        }

        public int getPrevHeight() {
            return prevHeight;
        }

        void setPrevHeight(int prevHeight) {
            this.prevHeight = prevHeight;
        }

        public int getNextHeight() {
            return nextHeight;
        }

        void setNextHeight(int newNextHeight) {
            nextHeight = newNextHeight;
        }

        public int getSleepBetween() {
            return sleepBetween;
        }

        void setSleepBetween(int newSleepBetween) {
            this.sleepBetween = newSleepBetween;
        }

        public long getPrevBalance() {
            return prevBalance;
        }

        void setPrevBalance(long newPrevBalance) {
            this.prevBalance = newPrevBalance;
        }

        public boolean getFreezeWhenSameBalance() {
            return freezeWhenSameBalance;
        }

        void setFreezeWhenSameBalance(boolean newFreezeWhenSameBalance) {
            this.freezeWhenSameBalance = newFreezeWhenSameBalance;
        }

        public long getMinActivationAmount() {
            return minActivationAmount;
        }

        void setMinActivationAmount(long newMinActivationAmount) {
            this.minActivationAmount = newMinActivationAmount;
        }
    }
}

File: src/main/kotlin/brs/at/AT.kt
/*
 * Some portion .. Copyright (c) 2014 CIYAM Developers

 Distributed under the MIT/X11 software license, please refer to the file LICENSE.txt
*/

package brs.at

import brs.db.BurstKey
import brs.entity.Block
import brs.entity.DependencyProvider
import brs.entity.Transaction
import brs.objects.Genesis
import brs.transaction.appendix.Appendix
import brs.transaction.appendix.Attachment
import brs.util.BurstException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

class AT : AtMachineState {
    val dbKey: BurstKey
    val name: String
    val description: String
    private val nextHeight: Int

    private constructor(
        dp: DependencyProvider,
        atId: Long,
        creator: Long,
        name: String,
        description: String,
        creationBytes: ByteArray,
        height: Int
    ) : super(dp, atId, creator, creationBytes, height) {
        this.name = name
        this.description = description
        dbKey = dp.atStore.atDbKeyFactory.newKey(atId)
        this.nextHeight = dp.blockchainService.height
    }

    constructor(
        dp: DependencyProvider, atId: Long, creator: Long, name: String, description: String, version: Short,
        stateBytes: ByteArray, csize: Int, dsize: Int, cUserStackBytes: Int, cCallStackBytes: Int,
        creationBlockHeight: Int, sleepBetween: Int, nextHeight: Int,
        freezeWhenSameBalance: Boolean, minActivationAmount: Long, apCode: ByteArray
    ) : super(
        dp, atId, creator, version,
        stateBytes, csize, dsize, cUserStackBytes, cCallStackBytes,
        creationBlockHeight, sleepBetween,
        freezeWhenSameBalance, minActivationAmount, apCode
    ) {
        this.name = name
        this.description = description
        dbKey = dp.atStore.atDbKeyFactory.newKey(atId)
        this.nextHeight = nextHeight
    }

    fun saveState() {
        var state: ATState? = dp.atStore.atStateTable[dp.atStore.atStateDbKeyFactory.newKey(id)]
        val prevHeight = dp.blockchainService.height
        val newNextHeight = prevHeight + waitForNumberOfBlocks
        if (state != null) {
            state.state = this.state
            state.prevHeight = prevHeight
            state.nextHeight = newNextHeight
            state.sleepBetween = this.sleepBetween
            state.prevBalance = getpBalance()
            state.freezeWhenSameBalance = freezeOnSameBalance()
            state.minActivationAmount = minActivationAmount()
        } else {
            state = ATState(
                dp, id,
                this.state, newNextHeight, sleepBetween,
                getpBalance(), freezeOnSameBalance(), minActivationAmount()
            )
        }
        dp.atStore.atStateTable.insert(state)
    }

    fun nextHeight(): Int {
        return nextHeight
    }

    open class ATState(
        dp: DependencyProvider,
        val atId: Long,
        var state: ByteArray,
        var nextHeight: Int,
        var sleepBetween: Int,
        var prevBalance: Long,
        var freezeWhenSameBalance: Boolean,
        var minActivationAmount: Long
    ) {
        val dbKey = dp.atStore.atStateDbKeyFactory.newKey(this.atId)
        var prevHeight: Int = 0
    }

    companion object {
        private val pendingFees = mutableMapOf<Long, Long>()
        private val pendingTransactions = mutableListOf<AtTransaction>()

        fun clearPendingFees() {
            pendingFees.clear()
        }

        fun clearPendingTransactions() {
            pendingTransactions.clear()
        }

        fun addPendingFee(id: Long, fee: Long) {
            pendingFees[id] = fee
        }

        fun addPendingFee(id: ByteArray, fee: Long) {
            addPendingFee(AtApiHelper.getLong(id), fee)
        }

        fun addPendingTransaction(atTransaction: AtTransaction) {
            pendingTransactions.add(atTransaction)
        }

        fun findPendingTransaction(recipientId: Long): Boolean {
            for (tx in pendingTransactions) {
                if (recipientId == tx.recipientId) {
                    return true
                }
            }
            return false
        }

        fun addAT(
            dp: DependencyProvider,
            atId: Long,
            creator: Long,
            name: String,
            description: String,
            creationBytes: ByteArray,
            height: Int
        ) {
            val at = AT(dp, atId, creator, name, description, creationBytes, height)

            dp.atController.resetMachine(at)

            dp.atStore.atTable.insert(at)

            at.saveState()

            val account = dp.accountService.getOrAddAccount(atId)
            account.apply(dp, ByteArray(32), height)
        }

        fun compressState(stateBytes: ByteArray?): ByteArray? {
            if (stateBytes == null || stateBytes.isEmpty()) {
                return null
            }

            ByteArrayOutputStream().use { bos ->
                GZIPOutputStream(bos).use { gzip ->
                    gzip.write(stateBytes)
                    gzip.flush()
                }
                return bos.toByteArray()
            }
        }

        fun decompressState(stateBytes: ByteArray?): ByteArray? {
            if (stateBytes == null || stateBytes.isEmpty()) {
                return null
            }

            ByteArrayInputStream(stateBytes).use { bis ->
                GZIPInputStream(bis).use { gzip ->
                    ByteArrayOutputStream().use { bos ->
                        gzip.copyTo(bos, 256)
                        return bos.toByteArray()
                    }
                }
            }
        }

        fun handleATBlockTransactionsListener(dp: DependencyProvider): (Block) -> Unit = { block ->
            pendingFees.forEach { (key, value) ->
                val atAccount = dp.accountService.getAccount(key)!!
                dp.accountService.addToBalanceAndUnconfirmedBalancePlanck(atAccount, -value)
            }

            val transactions = mutableListOf<Transaction>()
            for ((senderId, recipientId, amount, message) in pendingTransactions) {
                dp.accountService.addToBalanceAndUnconfirmedBalancePlanck(
                    dp.accountService.getAccount(senderId)!!,
                    -amount
                )
                dp.accountService.addToBalanceAndUnconfirmedBalancePlanck(
                    dp.accountService.getOrAddAccount(recipientId),
                    amount
                )

                val builder = Transaction.Builder(
                    dp, 1.toByte(), Genesis.creatorPublicKey,
                    amount, 0L, block.timestamp, 1440.toShort(), Attachment.AtPayment(dp)
                )

                builder.senderId(senderId)
                    .recipientId(recipientId)
                    .blockId(block.id)
                    .height(block.height)
                    .blockTimestamp(block.timestamp)
                    .ecBlockHeight(0)
                    .ecBlockId(0L)

                if (message != null) {
                    builder.message(Appendix.Message(dp, message, dp.blockchainService.height))
                }

                try {
                    val transaction = builder.build()
                    if (!dp.transactionDb.hasTransaction(transaction.id)) {
                        transactions.add(transaction)
                    }
                } catch (e: BurstException.NotValidException) {
                    throw Exception("Failed to construct AT payment transaction", e)
                }

            }

            if (transactions.isNotEmpty()) {
                // WATCH: Replace after transactions are converted!
                dp.transactionDb.saveTransactions(transactions)
            }
        }
    }
}
